File: AEDs2/TP/08/tp.java
import java.util.Scanner;

class java{
    private static int comparar(String texto1, String texto2, int i){
        int retorno = 0;
        if(i < texto1.length() && i < texto2.length()){
            if(texto1.charAt(i) != texto2.charAt(i)){
                retorno = 1;
            } else {
                retorno = retorno + comparar(texto1,texto2,i + 1);
            }
        } else if(texto1.length() == 0) retorno = 1;
        return retorno;
    }
    private static int comparar(String texto1, String texto2) { return comparar(texto1, texto2, 0); }
    
    private static boolean is_senha(String texto){
        boolean retorno = false;
        boolean a = false, b = false, c = false, d = false;
        for(int i = 0; i < texto.length(); i++){
            if(texto.charAt(i) >= 'a' && texto.charAt(i) <= 'z') a = true;
            else if(texto.charAt(i) >= 'A' && texto.charAt(i) <= 'Z') b = true;
            else if(texto.charAt(i) >= '0' && texto.charAt(i) <= '9') c = true;
            else d = true;
        }
        retorno = (a && b && c && d && texto.length() >= 8);
        return retorno;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String texto = "";

        texto = sc.nextLine();
        while(comparar(texto,"FIM") != 0){
            System.out.println((is_senha(texto) ? "SIM" : "NAO"));
            texto = sc.nextLine();
        }

        sc.close();
    }
}

File: AEDs2/TP/06/tp.c
#include <stdio.h>
#include <stdlib.h>

int comparar(const char *t1, const char *t2){
    while(*t1 && (*t1 == *t2)){
        t1 = t1 + 1;
        t2 = t2 + 1;
    }
    return *(unsigned char *)t1 - *(unsigned char *)t2;
}

int tamanho(char* texto){
    int n = 0;
    while(texto[n] != '\n' && texto[n] != '\0') n = n + 1;
    return n;
}

int anagrama(char* texto, int n){
    int retorno = 1;
    int alfabeto[26] = {0};
    int i = 0;
    int v = 0;
    int cont1 = 0, cont2 = 0;

    while(i < n){
        char c = texto[i];
        if(c != ' '){
            if(c >= 'A' && c <= 'Z') c = c + 32;
            if(c >= 'a' && c <= 'z'){
                int indice = c - 'a';
                if(!v){
                    alfabeto[indice]++;
                    cont1++;
                } else {
                    alfabeto[indice]--;
                    cont2++;
                }
            }
        } else v = 1;
        i = i + 1;
    }

    if(cont1 != cont2) retorno = 0;

    for(i = 0; i < 26; i++){
        if(alfabeto[i] != 0) retorno = 0;
    }

    return retorno;
}

int main(){
    char* linha = (char*) calloc(1000,sizeof(char));
    char fim[4] = "FIM";
    int n = 0;

    while(fgets(linha,1000,stdin) != NULL){
        n = tamanho(linha);
        linha[n] = '\0';
        if(comparar(linha,fim) == 0) break;
        printf("%s\n",(anagrama(linha,n) ? "SIM" : "NAO"));
    }

    free(linha);
    return 0;
}
